### ISSUE TITLE
[CT-3161] [Bug] incorporate datatypes from dbms extensions like postgis

### ISSUE THREAD

**Issue by jenna-jordan** [dbt-labs/dbt-core#8720]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

When codegen uses the datatypes coming from core to generate model yaml files, geometry data types are not correctly read. Instead, the output looks like:

```
      - name: geom_point_2249
        data_type: user-defined
        description: ""
```

### Expected Behavior

The full correct geometry datatype should be pass through instead:

```
      - name: geom_point_2249
        data_type: geometry(point, 2249)
        description: ""
```

### Steps To Reproduce

For a Postgres table with geometry types, run the macro `codegen.generate_model_yaml` - making sure it is codegen version 0.11.0 so the new parameter `include_data_types=True` is included.

### Environment

```markdown
- OS: Windows
- Python: 3.9.7
- dbt: 1.6
- codegen: 0.11.0
```

### Which database adapter are you using with dbt?

postgres

**Comment by dbeatty10** [comment]
Thanks for opening this @jenna-jordan !

Excited about your [upcoming talk](https://coalesce.getdbt.com/agenda/from-coast-to-coast-implementing-dbt-in-the-public-sector) with Ian and Laurie at Coalesce in San Diego next month 🤩

Are you hoping to use this in conjunction with an [enforced model contract](https://docs.getdbt.com/docs/collaborate/govern/model-contracts#how-to-define-a-contract)? Or are you just trying to generate the data types without needing or wanting to enforce them as a part of a model contract?

**Comment by jenna-jordan** [comment]
> Excited about your [upcoming talk](https://coalesce.getdbt.com/agenda/from-coast-to-coast-implementing-dbt-in-the-public-sector) with Ian and Laurie at Coalesce in San Diego next month 🤩

Thank you!!

> Are you hoping to use this in conjunction with an [enforced model contract](https://docs.getdbt.com/docs/collaborate/govern/model-contracts#how-to-define-a-contract)? Or are you just trying to generate the data types without needing or wanting to enforce them as a part of a model contract?

Yes - we plan on implementing enforced model contracts for pretty much all models. Our old method of transforms involved truncating rather than dropping a table specifically to preserve the schema, so model contracts get us the closest towards preserving this method. And we use geometry columns quite extensively in our database, with the type (point, polygon, etc) and SRID specified in the geometry type.

**Comment by dbeatty10** [comment]
Okay, if you're wanting this for enforcing a model contract, then we really have **two** things we'll need to solve:

1. codegen for user-defined types
2. enforcement for user-defined types

We've got some options right now to handle the codegen part. But after that, then I ran into the second issue when trying to actually enforce the contract 😭

### codegen for user-defined types

Do you want to try playing around with the following macro in your project? It's mostly taken from [`postgres__get_catalog`](https://github.com/dbt-labs/dbt-core/blob/a2d4424f922c2ba0f5884a5b39400b11901e165a/plugins/postgres/dbt/include/postgres/macros/catalog.sql) (which is a macro that that is used by `dbt docs generate`).

See options 1 and 2 depending on if you want to include "type modifiers" or not (e.g., `geometry(Point,4326)` vs. just `geometry`).

**tldr**: choose between one of these two:
```
format_type(col.atttypid, col.atttypmod)
```
or
```
format_type(col.atttypid, null)
```

The second option is what conforms to our current implementation of model contracts (which doesn't include type modifiers like size, scale, precision, etc).

The following is a proposed override of [`postgres__get_columns_in_relation`](https://github.com/dbt-labs/dbt-core/blob/a2d4424f922c2ba0f5884a5b39400b11901e165a/plugins/postgres/dbt/include/postgres/macros/adapters.sql#L61-L80):
```sql
{% macro postgres__get_columns_in_relation(relation) -%}
  {% call statement('get_columns_in_relation', fetch_result=True) %}

    select
      col.attname as column_name,

      -- Two options for formatting the data type:
      -- 1) include the type modifier when appliable (e.g., `geometry(Point,4326)`)
      -- 1) do not include the type modifier (e.g., `geometry`)

      -- Option 1:
      -- pg_catalog.format_type(col.atttypid, col.atttypmod) as data_type,

      -- Option 2:
      pg_catalog.format_type(col.atttypid, null) as data_type,

      null as character_maximum_length,
      null as numeric_precision,
      null as numeric_scale

    from pg_catalog.pg_namespace sch
    join pg_catalog.pg_class tbl on tbl.relnamespace = sch.oid
    join pg_catalog.pg_attribute col on col.attrelid = tbl.oid
    left outer join pg_catalog.pg_description tbl_desc on (tbl_desc.objoid = tbl.oid and tbl_desc.objsubid = 0)
    left outer join pg_catalog.pg_description col_desc on (col_desc.objoid = tbl.oid and col_desc.objsubid = col.attnum)

    where 1=1
      and not pg_is_other_temp_schema(sch.oid) -- not a temporary schema belonging to another session
      and tbl.relpersistence in ('p', 'u') -- [p]ermanent table or [u]nlogged table. Exclude [t]emporary tables
      and tbl.relkind in ('r', 'v', 'f', 'p') -- o[r]dinary table, [v]iew, [f]oreign table, [p]artitioned table. Other values are [i]ndex, [S]equence, [c]omposite type, [t]OAST table, [m]aterialized view
      and col.attnum > 0 -- negative numbers are used for system columns such as oid
      and not col.attisdropped -- column as not been dropped

      and tbl.relname = '{{ relation.identifier }}'
        {% if relation.schema %}
        and sch.nspname = '{{ relation.schema }}'
        {% endif %}

    order by
        sch.nspname,
        tbl.relname,
        col.attnum

  {% endcall %}
  {% set table = load_result('get_columns_in_relation').table %}
  {{ return(sql_convert_columns_in_relation(table)) }}
{% endmacro %}
```

**Comment by jenna-jordan** [comment]
So, when it comes to enforcing contracts based on the yaml generated by codegen, there are a couple of other datatype issues that I immediately ran into. One, that there is no character limit specified for varchar - it's just varchar rather than varchar(10). The second is with arrays - the generated datatype for a text array was just "array" - which caused an error when building the model, and I had to manually correct it to be "text array". Manually correcting the arrays is a pain but fine, because they aren't that common, whereas manually correcting all of the character limits would be a huge pain.

The only reason that we want the datatypes to be generated by codegen is so that we can use them to implement and enforce model contracts, so while it goes against current practice I would favor defaulting to the most specific datatype possible in all cases - geometries, varchars, numeric, etc.

However, that may be a larger issue to address separately from just being able to read a geometry datatype as geometry instead of user-defined - up to y'all on defining the scope!

**Comment by dbeatty10** [comment]
When enforcing model contracts, dbt does **not** compare granular details such as size, precision, or scale (["type modifiers"](https://www.postgresql.org/docs/current/sql-createtype.html) in Postgres parlance) -- it only compares the high-level type. If you need a more-precise assertion, our guidance is to [create/use a custom test](https://docs.getdbt.com/guides/best-practices/writing-custom-generic-tests).

[Excerpt from our docs](https://docs.getdbt.com/reference/resource-configs/contract#support):

> When dbt is comparing data types, it will not compare granular details such as size, precision, or scale. We don't think you should sweat the difference between varchar(256) and varchar(257), because it doesn't really affect the experience of downstream queriers. If you need a more-precise assertion, it's always possible to accomplish by [writing or using a custom test](https://docs.getdbt.com/guides/best-practices/writing-custom-generic-tests).

Did it work once you corrected the data type for the arrays?

If so, that's something that would be possible by changing the implementation of one or both of [these macros in `codegen`](https://github.com/dbt-labs/dbt-codegen/blob/af12570eb8618b71f22c8ad77593d25a33dacd26/macros/helpers/helpers.sql#L66-L80).

**Comment by nicods-fr** [comment]
Hello,
We prefer option 2 😇 : `pg_catalog.format_type(pg_attribute.atttypid, pg_attribute.atttypmod)`
I will read your contribution guide, maybe I can request à merge 🥷
